core-bundle/src/Resources/contao/library/Contao/Dbafs.php

```diff
@@ -494,9 +494,17 @@ public static function syncFiles()
 		}
 
 		$objDatabase = Database::getInstance();
+		$objDatabaseEngine = $objDatabase
+			->prepare("SELECT ENGINE FROM information_schema.TABLES WHERE TABLE_NAME = 'tl_files' AND TABLE_SCHEMA = ?")
+			->execute(Config::get('dbDatabase'));
+		$blnDatabaseSupportsTransaction = 1 === $objDatabaseEngine->numRows && 'InnoDB' === $objDatabaseEngine->fetchRow()[0];
 
-		// Lock the files table
-		$objDatabase->lockTables(array('tl_files'=>'WRITE'));
+		// Begin atomic database access
+		if($blnDatabaseSupportsTransaction) {
+			$objDatabase->beginTransaction();
+		} else {
+			$objDatabase->lockTables(array('tl_files'=>'WRITE'));
+		}
```

Comment:
Couldn’t we use both, table locking and transactions at the same time?

```php
$objDatabase->lockTables(array('tl_files'=>'WRITE'));
$objDatabase->beginTransaction();
// ...
$objDatabase->commitTransaction();
$objDatabase->unlockTables();
```

This way we wouldn not need to detect if transactions are supported at all and I think the performance gains should be similar?

Reply:
Nope, I tried that as well. Unfortunately as soon as table locking is in place the whole performance benefit goes away again. (See the sidenote my initial description.)

Reply:
@ausi @leofeyer 

I just tried this again in my new implementation (locking together with a transaction) and there it didn't make a huge difference. I'm not sure why this wasn't the case in my last tests here. I'm going to change this here - maybe anyone of you could run a test as well. I don't want run into a problem with database config intricacies :smile:.

Reply:
Tried this with different environments now and it seems to work everywhere so far.

Reply:
I can confirm that it also works on contao.org.